block-device-driver/src/lib.rs

```diff
@@ -1,42 +1,134 @@
-#![no_std]
+//! An abstraction of block devices.
+
+#![cfg_attr(not(test), no_std)]
+#![warn(missing_docs)]
 #![allow(async_fn_in_trait)]
 
+use elain::{Align, Alignment};
+
 /// A trait for a block devices
 ///
+/// [`BlockDevice<const SIZE: usize, const ALIGN: usize>`](BlockDevice) can be initialized with the following parameters.
+///
+/// - `SIZE`: The size of the block, this dictates the size of the internal buffer.
+/// - `ALIGN`: The alignment of the block buffers, this defaults to 1 unless
+///            specified in the trait impl. Alignment must be a power of two.
+///
+/// The generic parameter `SIZE` on [BlockDevice] is the number of _bytes_ in a block 
+/// for this block device.
+///
+/// All addresses are zero indexed, and the unit is blocks. For example to read bytes
+/// from 1024 to 1536, the supplied block address would be 2.
+///
 /// This trait can be implemented multiple times to support various different block sizes.
-pub trait BlockDevice<const SIZE: usize> {
+pub trait BlockDevice<const SIZE: usize, const ALIGN: usize = 1>
```

Comment:
maybe ALIGN should be an associated const? 

if you put it in the trait generics, a type can impl `BlockDevice<512, 1>` and `BlockDevice<512, 2>` and `BlockDevice<512, 4>` etc at the same time, which doesn't make much sense?

I see the alignment as an "output": "Does T implement BlockDevice for 512-byte blocks? Yes, and it requires alignment of 4". not "Does T implement BlockDevice for 512-byte blocks and 4-byte alignment? Yes/No"

This'll allow users of the trait like `BufStream` to automatically use the right alignment, instead of the user having to pass it in.

Reply:
This worked out much nicer, thanks for the tip! Hopefully, we can figure out a solution for `aligned`.